stdlib/src/builtin/sort.mojo

```diff
@@ -15,383 +15,175 @@
 
-    if cmp_fn[type](array[right], pivot_value):
-        right += 1
-    swap(array[end - 1], array[right])
-    return right
+    @always_inline
+    @parameter
+    fn _partition(low: Int, high: Int) -> Int:
+        var pivot = list[high]
+        var i = low - 1
+        for j in range(low, high):
+            if list[j] <= pivot:
+                i += 1
+                list[j], list[i] = list[i], list[j]
+        list[i + 1], list[high] = list[high], list[i + 1]
+        return i + 1
+
+    if low < high:
+        var pi = _partition(low, high)
+        _quick_sort(list, low, pi - 1)
+        _quick_sort(list, pi + 1, high)
 
 
 @always_inline
-fn _partition[
-    type: CollectionElement, cmp_fn: fn (type, type) capturing -> Bool
-](array: UnsafePointer[type], start: Int, end: Int) -> Int:
-    if start == end:
-        return end
-
-    var pivot = start + (end - start) // 2
-
-    var pivot_value = array[pivot]
-
-    var left = start
-    var right = end - 2
-
-    swap(array[pivot], array[end - 1])
-
-    while left < right:
-        if cmp_fn(array[left], pivot_value):
-            left += 1
-        elif not cmp_fn(array[right], pivot_value):
-            right -= 1
-        else:
-            swap(array[left], array[right])
-
-    if cmp_fn(array[right], pivot_value):
-        right += 1
-    swap(array[end - 1], array[right])
-    return right
-
-
-fn _estimate_initial_height(size: Int) -> Int:
-    # Compute the log2 of the size rounded upward.
-    var log2 = int((bitwidthof[DType.index]() - 1) ^ countl_zero(size | 1))
-    return max(2, log2)
-
-
-fn _quicksort[
-    type: AnyRegType, cmp_fn: _cmp_fn_type
-](array: Pointer[type], size: Int):
-    if size == 0:
-        return
-
-    var stack = List[Int](capacity=_estimate_initial_height(size))
-    stack.append(0)
-    stack.append(size)
-    while len(stack) > 0:
-        var end = stack.pop()
-        var start = stack.pop()
-
-        var len = end - start
-        if len < 2:
-            continue
-
-        if len == 2:
-            _small_sort[2, type, cmp_fn](array + start)
-            continue
-
-        if len == 3:
-            _small_sort[3, type, cmp_fn](array + start)
-            continue
-
-        if len == 4:
-            _small_sort[4, type, cmp_fn](array + start)
-            continue
-
-        if len == 5:
-            _small_sort[5, type, cmp_fn](array + start)
-            continue
-
-        if len < 32:
-            _insertion_sort[type, cmp_fn](array, start, end)
-            continue
-
-        var pivot = _partition[type, cmp_fn](array, start, end)
-
-        stack.append(pivot + 1)
-        stack.append(end)
-
-        stack.append(start)
-        stack.append(pivot)
+fn quick_sort[dtype: DType](inout list: List[Scalar[dtype]]):
+    """Sort list of scalars in-place with quick sort algorithm.
 
+    Parameters:
+        dtype: The dtype of the scalar.
 
-fn _quicksort[
-    type: CollectionElement, cmp_fn: fn (type, type) capturing -> Bool
```

Comment:
**Question** Previously, this quicksort function allowed specifying a comparator; we use this functionality internally for sorting by `greater_than` comparator for example instead of the default less than comparator.  Do you mind adding this level of flexibility and plumbing it through the various sorting algorithms?

Reply:
There a wrinkle with that, currently all sort algorithms are comparison based so it is not a problem. But Radix sort is not comparison based, hence it would mean we would have an issue to call it from the sort function, if sort function allows configurable comparison operator. I guess if the main motivation is for sorting in descending order, then we can introduce an ascending/descending parameter, instead of a sorting function. This should work for radix sort as well. If there are other motivations, then I guess we can go with operator and avoid calling radix sort from the sort function. We could still introduce it as a standalone function in the module though. 

Let me know what you prefer.

Reply:
That makes sense.  We have a couple of use cases internally for calling `quick_sort` with various comparators such as `>=`, `<=`, or `==` on the elements when doing the comparisons.  I don't think we should go the ascending/descending route since it's not much more work to be generic over any comparison function.  Perhaps we should add an explicit `sort_by` function that allows passing this generic functor working on comparison of two `ComparableCollectionElement` values?  Something like https://doc.rust-lang.org/std/primitive.slice.html#method.sort_by for example.  With this particularly, we can decide whether we want to push the function (and default it to a generic less than function for sorting purposes) in the `sort` API or have it be a different named function (such as the `sort_by`).

I think we should go in this general direction and then think about various things as follow-ups — do we want various sorting APIs or have a one-stop-shop like in C++ [here](https://en.cppreference.com/w/cpp/algorithm/ranges/sort)?  Of note, you mentioned non-comparison based sorting algorithms wouldn't be suitable for this sort of function signature.

What do you think is best?  Also CC @laszlokindrat @rparolin in case you have thoughts.

Reply:
The question is, how generic do we want the `sort` to be. 

The current internal functions signatures are:
`[type: AnyRegType, cmp_fn: _cmp_fn_type](array: Pointer[type])` 

and 

`[type: CollectionElement, cmp_fn: fn (type, type) capturing -> Bool](array: UnsafePointer[type])`

which are called from public signatures like:

`sort(inout buff: Pointer[Int], len: Int):`

`fn sort[type: DType](inout buff: Pointer[Scalar[type]], len: Int):`

`fn sort(inout v: List[Int]):`

`fn sort[type: DType](inout v: List[Scalar[type]]):`

and 

`fn sort[type: CollectionElement, cmp_fn: fn (type, type) capturing -> Bool](inout v: List[type]):`

The state of the sort module actually changed a lot since my PR was opened so I think it is best to take a step back. I will make this PR less aggressive I will just introduce a new `sort` function with following signature:

`fn sort[type: ComparableCollectionElement](inout v: List[type]):`

So the PR will introduce the initial concept of generic comparison sorting and then we can build upon that.